proposed/2022/pack-referenced-projects.md

```diff
@@ -0,0 +1,73 @@
+# Pack referenced projects
+
+- Author Name (https://github.com/maxkoshevoi)
+- Start Date (2022-03-28)
+- GitHub Issue (https://github.com/NuGet/Home/issues/3891)
+- GitHub PR (-)
+
+## Summary
+
+Allow referenced projects' artifacts to be packed inside the NuGet.
+
+## Motivation
+
+When SDK-style project is being packed as a NuGet, all projects that it references are considered as NuGet themselves that this NuGet depends on.
+This is not always a desired behavior, sometimes user needs artifacts from referenced project(s) to be packed into the NuGet.
+Currently there's no easy way to override this behavior except for ditching `csproj`, and  going back to `nuspec`, which becomes another source of truth and all that good stuff.
+
+## Explanation
+
+### Functional explanation
+
+Proposed solution is to take into account `IsPackable` property of the referenced projects. If it's `True` or not present (to preserve backward compatibility) - consider the project as a NuGet dependency, if it's `False` - pack its artifacts in.
+
+Second part is introduce new project-wide `PackReferencedProjects` property that would switch pack behavior for all `ProjectReference`s inside a single project.
```

Comment:
I would rather have it where each ``ProjectReference`` would instead have optional metadata ``Pack="true"`` or ``<Pack>true</Pack>`` inside it like how ``<None Include="..." Pack="true" ... />`` works.

Why? Because of unit test projects normally set themselves as non-packable.

Reply:
> Why? Because of unit test projects normally set themselves as non-packable. So with this it could be an option to include them in the package without being "package dependencies"

This is exactly how it would work with `IsPackable` approach. If project `N` references unit test project `UT`, `UT`s artifacts would be packed into `N`s NuGet since `UT` has `IsPackable` set to `False`. Am I missing something in your scenario?

> However this could be extended to not only just ProjectReference like PackageReference as well if one wanted to do that too with the metadata option.

This is interesting, but I don't see why would anybody want to pack NuGet inside another NuGet. Do you have any use case for that?

Reply:
Yes, for example I might have a package that depends on packages that are not on nuget.org (a different feed or is a version of the package that is not in nuget.org like packages in the dotnet-7 feed for example) and may want to explicitly tell those specific packages to be packed in as well so when my package deploys on nuget.org people do not complain about the missing dependency.

Reply:
Hm, makes sense, thanks.
Are you proposing adding `Pack="true"` to `ProjectReference` and that's it, or introducing `PackReferencedProjects` as well?
Or even maybe inferring default `ProjectReference.Pack` value based on `IsPackable`? 

All three might be to much for such feature, but `IsPackable` + `Pack` might be good. This way user don't need to add `Pack="true"` for packages that explicitly state that they are not packable.

Thinking about it, even just adding `ProjectReference.Pack` would be good MVP for this.

Reply:
I was thinking about only adding ``Pack`` to both ``ProjectReference`` and ``PackageReference``.

Reply:
Sounds good. Updated the proposal, could you take a look please?

Reply:
It looks good so far, thanks.